### ISSUE TITLE
[Question] Place Objects

### ISSUE THREAD

**Issue by maxmaynard** [Unity-Technologies/ml-agents#1657]
Hey all,

I'm creating a simulation where a ML agent places objects as far apart as possible from each other within an area. I'd like to do it with a variable number of objects, but I'm having some difficulties getting the agent to learn.

At the moment I'm using visual observations top down to try to get the brain to associate negative rewards with overlapping objects, and positive rewards for the amount of distance between each object and the next closest one, but it isn't learning like I'd hoped. 

Any thoughts?

**Comment by harperj** [comment]
Hi @maxmaynard, this seems like a reasonable experiment.  Unfortunately it's hard to provide much use here without seeing the specific environment, and we don't have sufficient resources to look into external environments at this time.

You might want to verify your reward is being set correctly and actions are behaving as expected.  If those are correct, you may find that you simply need more training steps; environments with visual observations can take some time to train.

Finally we do have some information about the [various hyperparameters available](https://github.com/Unity-Technologies/ml-agents/blob/master/docs/Training-ML-Agents.md#training-config-file).

I'll leave this discussion open in case others in the community can help further.

**Comment by maxmaynard** [comment]
Thanks for the tips. I've confirmed that rewards are being set properly with a player brain. The last large data set(does 300k steps count as large?) I tried I did have a mistake in the reward logic, so I'll try again. I've simplified to just providing a reward for placing the objects and a negative reward for overlapping. I'll see how that behaves over time.

**Comment by maxmaynard** [comment]
Oh, yeah, sure. The dimensions are 80x80 and the hyperparameters are below. I'm finishing up a training session that fits these parameters, but the data seems erratic overall. This is the cumulative reward graph over 800k steps with 11 training environments. There's also a screenshot of one environments; the orange dots are the objects being placed within the darker gray grid. I'd thought the contrast and simplicity would be enough that I could use a lower resolution than I've seen mentioned by other forum posts, but this is my first experiment in visual observations, so maybe I'm wrong.

<img width="473" alt="screen shot 2019-01-31 at 8 55 45 am" src="https://user-images.githubusercontent.com/21210011/52070587-055db600-2536-11e9-855e-fecfa30ca043.png">


<img width="1306" alt="screen shot 2019-01-31 at 8 53 15 am" src="https://user-images.githubusercontent.com/21210011/52070437-b283fe80-2535-11e9-91f7-b5758ebe70f1.png">



default:
    trainer: ppo
    batch_size: 1024
    beta: 5.0e-3
    buffer_size: 10240
    epsilon: 0.2
    gamma: 0.99
    hidden_units: 128
    lambd: 0.95
    learning_rate: 3.0e-4
    max_steps: 3.0e6
    memory_size: 256
    normalize: false
    num_epoch: 3
    num_layers: 2
    time_horizon: 64
    sequence_length: 64
    summary_freq: 1000
    use_recurrent: false
    use_curiosity: false
    curiosity_strength: 0.01
    curiosity_enc_size: 128

**Comment by maxmaynard** [comment]
In the simplified version, the brain receives a reward for placing a penny, and a penalty if they interact. The pennies themselves send the rewards to the brain via the agent; the overlap mechanics are using standard Unity triggers.

**Comment by MarcoMeter** [comment]
That are the rewards, but what does your agent as actions in fact?
Is it like placing two pennies at one step or does it have to push them around?

**Comment by maxmaynard** [comment]
Oh, every step it chooses a position within the board and places the penny there. It doesn't push it; just sets the transform position to the chosen spot.

**Comment by maxmaynard** [comment]
Given how the last action works, there's no guarantee it'll only take the number of steps equal to the number of pennies. There's a 50/50 chance a penny gets placed each step, as long as there is a penny to be placed. Otherwise it's done (once the last penny finishes it's checks).

But it could certainly be reduced to that.

**Comment by MarcoMeter** [comment]
I don't know if I get your environment right, but it sounds like that it is not the ideal use-case for Reinforcement Learning. Monte Carlo Tree Search would be more fitting, however it should be doable with Reinforcement Learning.

Regardless, if there are not that many possible states in the environment, I'd reduce the size of the hidden units.

**Comment by maxmaynard** [comment]
Hmm. I'll look into the Monte Carlo. In the meantime, I don't know what you're referring to when you say hidden units. 

Thank you for the advice by the way. I really appreciate it.

**Comment by MarcoMeter** [comment]
Hidden units in your hyperparameters.

**Comment by eshvk** [comment]
Thank you for this discussion, we are going to close this issue due to inactivity for now.
